### PR TITLE
Initializers are copied but separate from superclass initializers

### LINKED ISSUE #3373: [reactive-element] addInitializer adds initializer to superclass
### Which package(s) are affected?

Lit Core (lit / lit-html / lit-element / reactive-element)

### Description

While looking #3296 a bit more in depth, I realize that the culprit is not @lit-labs/context but `ReactiveElement.addInitializer`, which will add intializers to the superclass if it already contains its own initializers. 

[Documentation states](https://lit.dev/docs/components/lifecycle/#addInitializer): 

> Initializers are stored per-constructor. Adding an initializer to a subclass does not add it to a superclass. Since initializers are run in constructors, initializers will run in order of the class hierarchy, starting with superclasses and progressing to the instance's class.

This is not what happens as `addInitializer` only checks if this._initializer exists on the Class prototype chain. In case a superclass already has one, it will be added to it (the superclass). We should have a `hasOwnProperty` check there. 

```ts
   /*
   * Initializers are stored per-constructor. Adding an initializer to a
   * subclass does not add it to a superclass. Since initializers are run in
   * constructors, initializers will run in order of the class hierarchy,
   * starting with superclasses and progressing to the instance's class.
   *
   * @nocollapse
   */
  static addInitializer(initializer: Initializer) {
    this._initializers ??= [];
    this._initializers.push(initializer);
  }

```

### Reproduction

- go to [this playground](https://lit.dev/playground/#gist=410407cc4c6bb0051849d6633d0f0f5d)
- See that the base class has 2 initializers (while it shoul donly have one)

### Workaround

none

### Is this a regression?

No or unsure. This never worked, or I haven't tried before.

### Affected versions

Failing in 2.4.0

### Browser/OS/Node environment

NA

### PR DESCRIPTION
**Pull request by sorvell**
Fixes #3373.

### CODE REVIEW

**Comment by christophe-g** [comment]
>  and the lit team shall be in better position to efficiently tackle this (esp. on the good test requirement).

indeed, thanks !

**Review by kevinpschaaf — APPROVED** [review]
LGTM, add a changeset

File: packages/reactive-element/src/reactive-element.ts [hunk 1]
```diff
@@ -490,8 +490,9 @@ export abstract class ReactiveElement
    * @nocollapse
    */
   static addInitializer(initializer: Initializer) {
+    this.finalize();
     this._initializers ??= [];
-    this._initializers.push(initializer);
+    this._initializers!.push(initializer);
```
Comment (justinfagnani) [inline packages/reactive-element/src/reactive-element.ts:495]: 🏌️‍♂️: `(this._initializers ??= []).push(initializer)`

File: packages/reactive-element/src/reactive-element.ts [hunk 2]
```diff
@@ -762,6 +763,11 @@ export abstract class ReactiveElement
     // finalize any superclasses
     const superCtor = Object.getPrototypeOf(this) as typeof ReactiveElement;
     superCtor.finalize();
+    // Create own set of initializers for this class if any exist on the
+    // superclass and copy them down.
+    if (superCtor._initializers) {
+      this._initializers = [...superCtor._initializers];
```
Comment (justinfagnani) [inline packages/reactive-element/src/reactive-element.ts:769]: 🏌️‍♂️: `this._initializers = [...superCtor._initializers ?? []];`

**Review by justinfagnani — APPROVED** [review]

File: packages/reactive-element/src/reactive-element.ts (hunk 1 above)
Comment (sorvell) [reply packages/reactive-element/src/reactive-element.ts]: Done.

File: packages/reactive-element/src/reactive-element.ts (hunk 2 above)
Comment (sorvell) [reply packages/reactive-element/src/reactive-element.ts:769]: Done for small perf benefit. Added comment.

**Comment by christophe-g** [comment]
@sorvell, @justinfagnani - any chance to release a patch version for this ? Potential consequences of the bug can be dire (see https://github.com/lit/lit/issues/3296 - wich cause one of my app to break when using @lit-lab/context decorators)

**Pull request merged**